File: pilha/Pilha Estática/Java/Pilha Invertida/PilhaEstaticaInvertida.java
public class PilhaEstaticaInvertida implements Empilhavel {
	//variáveis de instância
	private int ponteiroTopo;
	private Object[] dados;
	
	//construtores
	public PilhaEstaticaInvertida() {
		this(10);
	}

	public PilhaEstaticaInvertida(int tamanho) {
		ponteiroTopo = tamanho;
		dados = new Object[tamanho];
	}
	
	@Override
	public void empilhar(Object dado) {
		if(!estaCheia()) {
			ponteiroTopo--;
			dados[ponteiroTopo] = dado;
		} else {
			System.err.println("Pilha Cheia!");
		}
	}

	@Override
	public Object desempilhar() {
		Object dadoTopo = null;
		if(!estaVazia()) {
			dadoTopo = dados[ponteiroTopo]; 
			ponteiroTopo++;
		} else {
			System.err.println("Pilha Vazia!");
		}
		return dadoTopo;
	}
	
	@Override
	public Object topo() {
		Object dadoTopo = null;
		if(!estaVazia()) {
			dadoTopo = dados[ponteiroTopo]; 
		} else {
			System.err.println("Pilha Vazia!");
		}
		return dadoTopo;	
	}
	
	@Override
	public boolean estaCheia() {
		return (ponteiroTopo == 0);
	}
	
	@Override
	public boolean estaVazia() {
		return (ponteiroTopo == dados.length);
	}
	
	@Override
	public String imprimir() {
		String resultado = "[";
		for(int i = ponteiroTopo; i <= dados.length-1; i++) {
			if (i == dados.length-1) {
				resultado += dados[i];
			}
			else {
				resultado += dados[i] +  ",";
			}
		}
		return resultado + "]";
	}
}

File: pilha/Pilha Estática/Kotlin/Pilha Invertida/PilhaEstaticaInvertida.kt
class PilhaEstaticaInvertida : Empilhavel {
    
	private var ponteiroTopo: Int
	private var dados: Array<Any?>
	
	constructor(tamanho: Int) {
		dados = arrayOfNulls(tamanho)
		ponteiroTopo = tamanho
	}
	
	constructor() : this(10)
    
	override fun atualizar(dado: Any?) {
		if (!estaVazia())
			dados[ponteiroTopo] = dado
		else
			println("Stack is empty!")
	}
		
	override fun empilhar(dado: Any?) {
		if (!estaCheia()) {
			ponteiroTopo--
			dados[ponteiroTopo] = dado
		} else {
			println("Stack is full!")
		}
	}
	
	override fun desempilhar(): Any? {
		var dadoTopo: Any? = null
		if (!estaVazia()) {
			dadoTopo = dados[ponteiroTopo]
			ponteiroTopo++
		} else {
			println("Stack is empty!")
		}
		return dadoTopo
	}
	
	override fun espiar(): Any? {
		var dadoTopo: Any? = null
		if (!estaVazia()) {
			dadoTopo = dados[ponteiroTopo]
		} else {
			println("Stack is empty!")
		}
		return dadoTopo
	}
	
	override fun estaCheia(): Boolean {
		return (ponteiroTopo == 0)
	}
	
	override fun estaVazia(): Boolean {
		return (ponteiroTopo == dados.size)
	}
	
	override fun imprimir(): String {
		var resultado = "["
		for (i in ponteiroTopo .. dados.size-1) {
			if (i == dados.size-1)
				resultado += "${dados[i]}"
			else
				resultado += "${dados[i]}, "
		}
		return "$resultado]"
	}
}